onnxruntime/contrib_ops/cpu/signal/window_functions.cc

```diff
@@ -0,0 +1,325 @@
+#include <functional>
+
+#include "core/platform/threadpool.h"
+
+#include <complex>
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    HannWindow,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    HannWindow);
+
+ONNX_OPERATOR_KERNEL_EX(
+    HammingWindow,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    HammingWindow);
+
+ONNX_OPERATOR_KERNEL_EX(
+    BlackmanWindow,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    BlackmanWindow);
+
+
+ONNX_OPERATOR_KERNEL_EX(
+    MelWeightMatrix,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float>())
+        .TypeConstraint("T3", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    MelWeightMatrix);
+
+
+template <typename T>
+static Status cosine_sum_window(Tensor* Y, size_t size, float a0, float a1, float a2) {
+  auto* Y_data = reinterpret_cast<T*>(Y->MutableDataRaw());
+
+  // Calculate the radians to increment per sample
+  constexpr double pi = 3.14159265;
+  constexpr double tau = 2 * pi;
+  const double angular_increment = tau / size;
+
+  for (size_t i = 0; i < size; i++) {
+    auto a2_component = a2 == 0 ? 0 : (a2 * cos(2 * angular_increment * i));
+
+    T& value = *(Y_data + i);
+    value = static_cast<T>(a0 - (a1 * cos(angular_increment * i)) + a2_component);
+  }
+
+  return Status::OK();
+}
+
+template <typename T>
+static T get_scalar_value_from_tensor(const Tensor* tensor) {
+  ORT_ENFORCE(tensor->Shape().Size() == 1, "ratio input should have a single value.");
+  auto data_type = tensor->DataType()->AsPrimitiveDataType()->GetDataType();
+  switch (data_type) {
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+      return static_cast<T>(*reinterpret_cast<const float*>(tensor->DataRaw()));
+    case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:
+      return static_cast<T>(*reinterpret_cast<const double*>(tensor->DataRaw()));
+    case ONNX_NAMESPACE::TensorProto_DataType_INT32:
```

Comment:
where's int8, int16, float16 and the unsigned versions?

Reply:
This function is actually only used for reading scale attribute inputs which take on a limited range of values.